File: src/main/kotlin/com/stocks/service/AssetService.kt
package com.stocks.service

import com.stocks.dto.AssetRequest
import com.stocks.dto.AssetResponse
import com.stocks.dto.toResponse
import com.stocks.model.AssetEntity
import com.stocks.model.TransactionEntity
import com.stocks.model.Transactions
import org.jetbrains.exposed.v1.core.eq
import org.jetbrains.exposed.v1.jdbc.transactions.transaction
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.web.server.ResponseStatusException

@Service
class AssetService(
    private val calculationService: CalculationService,
) {
    fun findAll(): List<AssetResponse> =
        transaction {
            AssetEntity.all().sortedBy { it.ticker.value }.map { it.toResponse() }
        }

    fun findFiltered(
        type: String?,
        position: String?,
        delisted: String? = null,
    ): List<AssetResponse> =
        transaction {
            var assets = AssetEntity.all().sortedBy { it.ticker.value }.toList()

            if (!type.isNullOrBlank()) {
                assets = assets.filter { it.type == type }
            }

            if (!position.isNullOrBlank() && position != "all") {
                assets =
                    when (position) {
                        "with" -> assets.filter { it.hasPosition }
                        "without" -> assets.filter { !it.hasPosition }
                        else -> assets
                    }
            }

            if (!delisted.isNullOrBlank()) {
                assets =
                    when (delisted) {
                        "active" -> assets.filter { !it.delisted }
                        "delisted" -> assets.filter { it.delisted }
                        else -> assets
                    }
            }

            assets.map { it.toResponse() }
        }

    fun computeTickersWithPosition(): Set<String> =
        transaction {
            AssetEntity
                .all()
                .toList()
                .filter { it.hasPosition }
                .map { it.ticker.value }
                .toSet()
        }

    fun findById(ticker: String): AssetResponse? =
        transaction {
            AssetEntity.findById(ticker.uppercase())?.toResponse()
        }

    fun create(request: AssetRequest): AssetResponse {
        request.validate()
        val normalizedTicker = request.normalizedTicker()

        return transaction {
            if (AssetEntity.findById(normalizedTicker) != null) {
                throw ResponseStatusException(HttpStatus.CONFLICT, "Ticker already registered")
            }

            AssetEntity
                .new(normalizedTicker) {
                    yfTicker = request.yfTicker
                    name = request.name
                    type = request.type ?: "STOCK"
                    currency = request.currency
                }.toResponse()
        }
    }

    fun update(
        ticker: String,
        name: String?,
        type: String?,
        yfTicker: String?,
        currency: String?,
        delisted: Boolean,
    ): AssetResponse =
        transaction {
            val asset =
                AssetEntity.findById(ticker.uppercase())
                    ?: throw ResponseStatusException(HttpStatus.NOT_FOUND, "Asset not found")

            asset.name = name?.ifBlank { null }
            asset.type = type?.ifBlank { "STOCK" }
            asset.yfTicker = yfTicker?.ifBlank { null }
            asset.currency = currency?.ifBlank { "BRL" } ?: asset.currency
            asset.delisted = delisted

            asset.toResponse()
        }

    fun delete(ticker: String) =
        transaction {
            val asset =
                AssetEntity.findById(ticker.uppercase())
                    ?: throw ResponseStatusException(HttpStatus.NOT_FOUND, "Asset not found")
            asset.delete()
        }

    fun exists(ticker: String): Boolean =
        transaction {
            AssetEntity.findById(ticker.uppercase()) != null
        }

    fun refreshPositionFields(assetId: String) {
        transaction {
            val asset =
                AssetEntity.findById(assetId.uppercase()) ?: return@transaction

            val txs =
                TransactionEntity
                    .find { Transactions.assetId eq assetId.uppercase() }
                    .toList()

            if (txs.isEmpty()) {
                asset.hasPosition = false
                asset.quantity = 0.0
                asset.avgPrice = 0.0
                asset.avgPriceBrl = 0.0
                asset.totalCost = 0.0
                asset.totalCostBrl = 0.0
                asset.realizedPnl = 0.0
                asset.realizedPnlBrl = 0.0
                return@transaction
            }

            val data =
                txs.map {
                    TransactionData(
                        type = it.type,
                        quantity = it.quantity,
                        price = it.price,
                        fees = it.fees,
                        date = it.date,
                        priceBrl = it.priceBrl,
                        feesBrl = it.feesBrl,
                    )
                }

            val calc = calculationService.calculatePosition(data)
            asset.hasPosition = calc.quantity > 0
            asset.quantity = calc.quantity
            asset.avgPrice = calc.avgPrice
            asset.avgPriceBrl = calc.avgPriceBrl
            asset.totalCost = calc.totalCost
            asset.totalCostBrl = calc.totalCostBrl
            asset.realizedPnl = calc.realizedPnl
            asset.realizedPnlBrl = calc.realizedPnlBrl
        }
    }

    fun refreshAllPositionFields() {
        transaction {
            val assets = AssetEntity.all().toList()
            for (asset in assets) {
                val txs = asset.transactions.toList()

                if (txs.isEmpty()) {
                    asset.hasPosition = false
                    asset.quantity = 0.0
                    asset.avgPrice = 0.0
                    asset.avgPriceBrl = 0.0
                    asset.totalCost = 0.0
                    asset.totalCostBrl = 0.0
                    asset.realizedPnl = 0.0
                    asset.realizedPnlBrl = 0.0
                    continue
                }

                val data =
                    txs.map {
                        TransactionData(
                            type = it.type,
                            quantity = it.quantity,
                            price = it.price,
                            fees = it.fees,
                            date = it.date,
                            priceBrl = it.priceBrl,
                            feesBrl = it.feesBrl,
                        )
                    }

                val calc = calculationService.calculatePosition(data)
                asset.hasPosition = calc.quantity > 0
                asset.quantity = calc.quantity
                asset.avgPrice = calc.avgPrice
                asset.avgPriceBrl = calc.avgPriceBrl
                asset.totalCost = calc.totalCost
                asset.totalCostBrl = calc.totalCostBrl
                asset.realizedPnl = calc.realizedPnl
                asset.realizedPnlBrl = calc.realizedPnlBrl
            }
        }
    }
}

File: src/modules/asset/asset.service.ts
import type { Db } from '../../config/db.js'
import { calculatePosition, type TransactionData } from '../../domain/calculation.js'
import type { TransactionType } from '../../domain/constants.js'
import type { Transaction } from '../../generated/prisma/client.js'
import { HttpError } from '../../shared/http-error.js'
import type { IsoDate } from '../../shared/iso-date.js'
import type { AssetClassService } from '../allocation/asset-class.service.js'
import {
  type AssetFilters,
  type AssetView,
  type ParsedAssetRequest,
  toAssetView,
} from './asset.schema.js'

/** Porte de src/main/kotlin/com/stocks/service/AssetService.kt. */
export class AssetService {
  constructor(
    private readonly db: Db,
    private readonly assetClasses: AssetClassService,
  ) {}

  async findAll(): Promise<AssetView[]> {
    const assets = await this.db.asset.findMany({ orderBy: { ticker: 'asc' } })
    return assets.map(toAssetView)
  }

  async findFiltered(filters: AssetFilters): Promise<AssetView[]> {
    const { type, position, delisted } = filters

    const assets = await this.db.asset.findMany({
      where: {
        ...(type ? { type } : {}),
        ...(position === 'with' ? { hasPosition: true } : {}),
        ...(position === 'without' ? { hasPosition: false } : {}),
        ...(delisted === 'active' ? { delisted: false } : {}),
        ...(delisted === 'delisted' ? { delisted: true } : {}),
      },
      orderBy: { ticker: 'asc' },
    })
    return assets.map(toAssetView)
  }

  async computeTickersWithPosition(): Promise<Set<string>> {
    const assets = await this.db.asset.findMany({
      where: { hasPosition: true },
      select: { ticker: true },
    })
    return new Set(assets.map((a) => a.ticker))
  }

  async findById(ticker: string): Promise<AssetView | null> {
    const asset = await this.db.asset.findUnique({ where: { ticker: ticker.toUpperCase() } })
    return asset === null ? null : toAssetView(asset)
  }

  async exists(ticker: string): Promise<boolean> {
    const count = await this.db.asset.count({ where: { ticker: ticker.toUpperCase() } })
    return count > 0
  }

  async create(request: ParsedAssetRequest): Promise<AssetView> {
    if (await this.exists(request.ticker)) {
      throw HttpError.conflict('Ticker already registered')
    }
    const type = request.type ?? 'STOCK'
    const asset = await this.db.asset.create({
      data: {
        ticker: request.ticker,
        yfTicker: request.yfTicker ?? null,
        name: request.name ?? null,
        type,
        currency: request.currency,
        // Nasce classificado pelo tipo; a tela de alocação corrige o que ficar torto.
        assetClassId: await this.assetClasses.defaultClassIdForType(type),
      },
    })
    return toAssetView(asset)
  }

  async update(
    ticker: string,
    fields: {
      name?: string | null
      type?: string | null
      yfTicker?: string | null
      currency?: string | null
      delisted: boolean
    },
  ): Promise<AssetView> {
    const normalized = ticker.toUpperCase()
    const current = await this.db.asset.findUnique({ where: { ticker: normalized } })
    if (current === null) throw HttpError.notFound('Asset not found')

    const asset = await this.db.asset.update({
      where: { ticker: normalized },
      data: {
        name: blankToNull(fields.name),
        type: fields.type || current.type,
        yfTicker: blankToNull(fields.yfTicker),
        currency: fields.currency ? fields.currency : current.currency,
        delisted: fields.delisted,
      },
    })
    return toAssetView(asset)
  }

  async delete(ticker: string): Promise<void> {
    const normalized = ticker.toUpperCase()
    if (!(await this.exists(normalized))) throw HttpError.notFound('Asset not found')
    // As transações, proventos, preços e snapshots saem por ON DELETE CASCADE.
    await this.db.asset.delete({ where: { ticker: normalized } })
  }

  /** Recalcula quantidade, preço médio e resultado realizado de um ativo. */
  async refreshPositionFields(assetId: string): Promise<void> {
    const normalized = assetId.toUpperCase()
    const asset = await this.db.asset.findUnique({ where: { ticker: normalized } })
    if (asset === null) return

    const transactions = await this.db.transaction.findMany({ where: { assetId: normalized } })
    await this.db.asset.update({
      where: { ticker: normalized },
      data: positionFields(transactions),
    })
  }

  /**
   * Recalcula todos os ativos.
   *
   * O Kotlin percorria os ativos acessando `asset.transactions`, o que dispara uma query
   * por ativo (N+1). Aqui são duas queries, e os updates vão numa transação só.
   */
  async refreshAllPositionFields(): Promise<void> {
    const [assets, transactions] = await Promise.all([
      this.db.asset.findMany({ select: { ticker: true } }),
      this.db.transaction.findMany(),
    ])

    const byAsset = new Map<string, Transaction[]>()
    for (const tx of transactions) {
      const list = byAsset.get(tx.assetId)
      if (list === undefined) byAsset.set(tx.assetId, [tx])
      else list.push(tx)
    }

    await this.db.$transaction(
      assets.map((asset) =>
        this.db.asset.update({
          where: { ticker: asset.ticker },
          data: positionFields(byAsset.get(asset.ticker) ?? []),
        }),
      ),
    )
  }
}

/** Campos de posição derivados das transações. Lista vazia zera tudo. */
function positionFields(transactions: readonly Transaction[]) {
  if (transactions.length === 0) {
    return {
      hasPosition: false,
      quantity: 0,
      avgPrice: 0,
      avgPriceBrl: 0,
      totalCost: 0,
      totalCostBrl: 0,
      realizedPnl: 0,
      realizedPnlBrl: 0,
    }
  }

  const calc = calculatePosition(transactions.map(toTransactionData))
  return {
    hasPosition: calc.quantity > 0,
    quantity: calc.quantity,
    avgPrice: calc.avgPrice,
    avgPriceBrl: calc.avgPriceBrl,
    totalCost: calc.totalCost,
    totalCostBrl: calc.totalCostBrl,
    realizedPnl: calc.realizedPnl,
    realizedPnlBrl: calc.realizedPnlBrl,
  }
}

export function toTransactionData(tx: Transaction): TransactionData {
  return {
    type: tx.type as TransactionType,
    quantity: tx.quantity,
    price: tx.price,
    fees: tx.fees,
    date: tx.date as IsoDate,
    priceBrl: tx.priceBrl,
    feesBrl: tx.feesBrl,
  }
}

function blankToNull(value: string | null | undefined): string | null {
  return value === undefined || value === null || value === '' ? null : value
}
